### PR TITLE
TextDecoder's Decode now receives a BufferSource as input

### LINKED ISSUE #20344: Use BufferSource in TextDecoder API
Now that #20267 has merged, we can uncomment the `decode` method in TextDecoder.webidl that accepts BufferSource and remove the one that fakes it by accepting an `option` value. This will require similar cleanup as #20342.

Code: `components/script/dom/webidls/TextDecoder.webidl`, `components/script/dom/textdecoder.rs`
Tests: `./mach test-wpt tests/wpt/web-platform-tests/encoding/textdecoder*`

### PR DESCRIPTION
**Pull request by pvdrz**
r? jdm

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #20344 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20413)

### CODE REVIEW

File: components/script/dom/textdecoder.rs [hunk 1]
```diff
@@ -36,21 +36,21 @@ impl TextDecoder {
 
     pub fn new(global: &GlobalScope, encoding: &'static Encoding, fatal: bool) -> DomRoot<TextDecoder> {
         reflect_dom_object(Box::new(TextDecoder::new_inherited(encoding, fatal)),
-                           global,
-                           TextDecoderBinding::Wrap)
+        global,
+        TextDecoderBinding::Wrap)
```
Comment (jdm) [inline components/script/dom/textdecoder.rs]: nit: please revert this indentation change.

File: components/script/dom/textdecoder.rs [hunk 2]
```diff
@@ -36,21 +36,21 @@ impl TextDecoder {
...
-        };
-        Ok(TextDecoder::new(global, encoding, options.fatal))
-    }
+        -> Fallible<DomRoot<TextDecoder>> {
+            let encoding = match Encoding::for_label_no_replacement(label.as_bytes()) {
+                None => return TextDecoder::make_range_error(),
+                Some(enc) => enc
+            };
+            Ok(TextDecoder::new(global, encoding, options.fatal))
+        }
```
Comment (jdm) [inline components/script/dom/textdecoder.rs]: nit: revert these formatting changes.

File: components/script/dom/textdecoder.rs [hunk 3]
```diff
@@ -65,32 +65,31 @@ impl TextDecoderMethods for TextDecoder {
...
-            Ok(data) => data,
-            Err(_)   => {
-                return Err(Error::Type("Argument to TextDecoder.decode is not an ArrayBufferView".to_owned()));
-            }
-        };
+    fn Decode(
+        &self,
+        input: Option<ArrayBufferViewOrArrayBuffer>,
+        _options: &TextDecodeOptions)
+        -> Fallible<USVString> {
```
Comment (jdm) [inline components/script/dom/textdecoder.rs]: nit: please format this as:
```rust
    _options: &TextDecodeOptions
) -> Fallible<USVString> {
    match input {
```

**Review by jdm — CHANGES_REQUESTED** [review]
Only small formatting nits :)

File: components/script/dom/textdecoder.rs [hunk 4]
```diff
@@ -65,32 +65,31 @@ impl TextDecoderMethods for TextDecoder {
...
-            Ok(data) => data,
-            Err(_)   => {
-                return Err(Error::Type("Argument to TextDecoder.decode is not an ArrayBufferView".to_owned()));
-            }
-        };
+    fn Decode(
+        &self,
+        input: Option<ArrayBufferViewOrArrayBuffer>,
+        _options: &TextDecodeOptions
+        ) -> Fallible<USVString> {
```
Comment (jdm) [inline components/script/dom/textdecoder.rs]: nit: unindent this line by 4 spaces, and the rest of the function as well.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
An ini file need to be [updated](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations):
```
  ▶ Unexpected subtest result in /encoding/api-basics.html:
  └ PASS [expected FAIL] Encode/decode round trip: utf-8

  ▶ Unexpected subtest result in /encoding/api-basics.html:
  └ PASS [expected FAIL] Decode sample: utf-16le

  ▶ Unexpected subtest result in /encoding/api-basics.html:
  └ PASS [expected FAIL] Decode sample: utf-16be

  ▶ Unexpected subtest result in /encoding/api-basics.html:
  └ PASS [expected FAIL] Decode sample: utf-16
```

**Comment by christianpoveda** [comment]
Sorry, fixed it

**Pull request merged**